flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/StreamTableEnvironment.scala

```diff
@@ -234,11 +234,12 @@ abstract class StreamTableEnvironment(
             "UpsertStreamTableSink requires that Table has a full primary keys if it is updated.")
         }
         val outputType = sink.getOutputType
+        val resultType = getResultType(table.getRelNode, optimizedPlan)
```

Comment:
Shall we consider assembling the `resultType` directly from the `sink` parameter?

Reply:
Do you mean adding method such as `getResultType` in `TableSink`? The `resultType`  means the type of the result of the current sql query. IMO, even we can specify the `resultType` in `TableSink`, it can only be used for validation purpose, we should not use it directly. Thoughts?

Reply:
Since the `fieldNames` and `fieldTypes` in parameter `sink` are generated based on the raw RowType, 
https://github.com/apache/flink/blob/81dc260dc653085b9dbf098e8fd70a72d2d0828e/flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala#L848 while the `resultType` here is generated from the optimized plan. I just wonder whether they'll be inconsistent.

Reply:
The `resultType` generated from the optimized plan contains time indicator information. In `StreamTableEnvironment.translate`, it needs this information to transform the time indicator column to `TimeStamp` type. For the type consistent issue, it will be validated during converting CRow to output type:
https://github.com/apache/flink/blob/81dc260dc653085b9dbf098e8fd70a72d2d0828e/flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala#L941

Reply:
Yes, you are right. Thanks for the explanation @dianfu.